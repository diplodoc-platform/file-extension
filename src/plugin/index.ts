export type {PluginOptions, TransformOptions} from './transform';
export {transform} from './transform';
export {
    FileClassName,
    FileSpecialAttr,
    LinkHtmlAttr as FileHtmlAttr,
    FILE_TOKEN,
    FILE_TO_LINK_ATTRS_MAP,
    RULE_NAME as FILE_RULE_NAME,
    KNOWN_ATTRS as FILE_KNOWN_ATTRS,
    REQUIRED_ATTRS as FILE_REQUIRED_ATTRS,
    PREFIX as FILE_MARKUP_PREFIX,
} from './const';
