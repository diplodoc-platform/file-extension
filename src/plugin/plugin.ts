import type MarkdownIt from 'markdown-it';

import {
    ENV_FLAG_NAME,
    FILE_TOKEN,
    FILE_TO_LINK_ATTRS_MAP,
    FileClassName,
    FileSpecialAttr,
    KNOWN_ATTRS,
    PREFIX,
    PREFIX_LENGTH,
    REQUIRED_ATTRS,
    RULE_NAME,
} from './const';

export type FileOptions = {
    fileExtraAttrs?: [string, string][];
};

export const filePlugin: MarkdownIt.PluginWithOptions<FileOptions> = (md, opts) => {
    md.inline.ruler.push(RULE_NAME, (state, silent) => {
        if (state.src.substring(state.pos, state.pos + PREFIX_LENGTH) !== PREFIX) {
            return false;
        }

        // the rest of line after '{% file '
        const searchStr = state.src.slice(state.pos + PREFIX_LENGTH, state.posMax);
        // loking for pattern 'src="..." name="..." etc="value" %}'
        const matchResult = searchStr.match(/^((?:\s*\w+=(?:"[^"]+"|'[^']+')\s)+)\s*%}/);
        if (!matchResult) {
            return false;
        }

        const paramsGroupLength = matchResult[0].length; // '(src="..." name="...")'.length
        const paramsStr = matchResult[1]; // 'src="..." name="..."'

        // find pairs of key="foo" or key='bar'
        const params = paramsStr.match(/\w+=(?:"[^"]+"|'[^']+')/g);
        if (!params) {
            return false;
        }

        const attrsObj: Record<string, string> = {};
        params.forEach((param) => {
            const indexKey = param.indexOf('=');
            const key = param.slice(0, indexKey);
            const value = param.slice(indexKey + 2, -1);
            if (KNOWN_ATTRS.includes(key) && value) {
                attrsObj[key] = value;
            }
        });

        const hasAllRequiredAttrs = REQUIRED_ATTRS.every((attr) => attr in attrsObj);
        if (!hasAllRequiredAttrs) {
            return false;
        }

        if (!silent) {
            const token = state.push(FILE_TOKEN, '', 0);
            token.block = false;
            token.markup = PREFIX;
            token.content = attrsObj[FileSpecialAttr.Name];
            token.attrs = Object.entries(attrsObj);
            token.attrSet('class', FileClassName.Link);

            for (const attr of token.attrs) {
                if (attr[0] in FILE_TO_LINK_ATTRS_MAP) {
                    attr[0] = FILE_TO_LINK_ATTRS_MAP[attr[0] as FileSpecialAttr];
                }
            }

            if (Array.isArray(opts?.fileExtraAttrs)) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                token.attrs.push(...opts!.fileExtraAttrs);
            }
        }

        state.pos = state.pos + PREFIX_LENGTH + paramsGroupLength;

        state.env ??= {};
        state.env[ENV_FLAG_NAME] = true;

        return true;
    });

    md.renderer.rules[FILE_TOKEN] = (tokens, idx, _opts, _env, self) => {
        const token = tokens[idx];
        const iconHtml = `<span class="${md.utils.escapeHtml(FileClassName.Icon)}"></span>`;
        return `<a${self.renderAttrs(token)}>${iconHtml}${md.utils.escapeHtml(token.content)}</a>`;
    };
};
