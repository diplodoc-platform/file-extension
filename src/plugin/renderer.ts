import {FILE_TOKEN, FileClassName} from './const';

export const fileRenderer: markdownit.PluginSimple = (md) => {
    md.renderer.rules[FILE_TOKEN] = (tokens, idx, _opts, _env, self) => {
        const token = tokens[idx];
        const iconHtml = `<span class="${md.utils.escapeHtml(FileClassName.Icon)}"></span>`;
        return `<a${self.renderAttrs(token)}>${iconHtml}${md.utils.escapeHtml(token.content)}</a>`;
    };
};
