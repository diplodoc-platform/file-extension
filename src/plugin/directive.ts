import type {FileOptions} from './plugin';

import {directiveParser, registerInlineDirective} from '@diplodoc/directive';

import {fileRenderer} from './renderer';
import {ENV_FLAG_NAME, FILE_TOKEN, FileClassName, LinkHtmlAttr} from './const';

const ALLOWED_ATTRS: readonly string[] = [
    LinkHtmlAttr.ReferrerPolicy,
    LinkHtmlAttr.Rel,
    LinkHtmlAttr.Target,
    LinkHtmlAttr.Type,
    LinkHtmlAttr.HrefLang,
];

export const fileDirective: markdownit.PluginWithOptions<FileOptions> = (md, opts = {}) => {
    const {fileExtraAttrs} = opts;

    fileRenderer(md);

    md.use(directiveParser());

    registerInlineDirective(md, 'file', (state, params) => {
        if (!params.content || !params.dests) {
            return false;
        }

        const filename = params.content.raw;
        const filelink = params.dests.link || '';

        const token = state.push(FILE_TOKEN, '', 0);
        token.block = false;
        token.markup = ':file';
        token.content = params.content.raw;

        token.attrSet('class', FileClassName.Link);
        token.attrSet(LinkHtmlAttr.Href, filelink);
        token.attrSet(LinkHtmlAttr.Download, filename);

        if (params.attrs) {
            for (const attrName of ALLOWED_ATTRS) {
                if (params.attrs[attrName]) {
                    token.attrSet(attrName, params.attrs[attrName]);
                }
            }
        }

        if (Array.isArray(fileExtraAttrs)) {
            for (const [name, value] of fileExtraAttrs) {
                token.attrSet(name, value);
            }
        }

        state.env ??= {};
        state.env[ENV_FLAG_NAME] = true;

        return true;
    });
};
