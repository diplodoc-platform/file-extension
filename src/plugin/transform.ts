import MarkdownIt from 'markdown-it';

import {type FileOptions, filePlugin} from './plugin';
import {ENV_FLAG_NAME} from './const';
import {hidden} from './utils';
import {fileDirective} from './directive';

export type PluginOptions = FileOptions & {
    output?: string;
};

export type RuntimeObj = {style: string};

export type TransformOptions = {
    runtime?: string | {style: string};
    bundle?: boolean;
    extraAttrs?: FileOptions['fileExtraAttrs'];
    /**
     * Enables directive syntax of yfm-file.
     *
     * - 'disabled' – directive syntax is disabled.
     * - 'enabled' – directive syntax is enabled; old (curly bracket) syntax is also enabled.
     * - 'only' – enabled only directive syntax; old (curly bracket) syntax is disabled.
     *
     * @default 'disabled'
     */
    directiveSyntax?: 'disabled' | 'enabled' | 'only';
    /** @internal */
    onBundle?: (env: {bundled: Set<string>}, output: string, runtime: RuntimeObj) => void;
};

const registerTransform = (
    md: MarkdownIt,
    {
        extraAttrs,
        bundle,
        onBundle,
        runtime,
        output,
        directiveSyntax,
    }: Pick<TransformOptions, 'extraAttrs' | 'onBundle'> & {
        bundle: boolean;
        runtime: {style: string};
        output: string;
        directiveSyntax: NonNullable<TransformOptions['directiveSyntax']>;
    },
) => {
    if (directiveSyntax === 'disabled' || directiveSyntax === 'enabled') {
        filePlugin(md, {fileExtraAttrs: extraAttrs});
    }
    if (directiveSyntax === 'enabled' || directiveSyntax === 'only') {
        fileDirective(md, {fileExtraAttrs: extraAttrs});
    }

    md.core.ruler.push('yfm_file_after', ({env}) => {
        if (env?.[ENV_FLAG_NAME]) {
            env.meta ||= {};
            env.meta.style ||= [];
            env.meta.style.push(runtime.style);

            if (bundle && onBundle) {
                hidden(env, 'bundled', new Set<string>());
                onBundle(env, output, runtime);
            }
        }
    });
};

export const transform = (opts: TransformOptions = {}) => {
    const {bundle = true, directiveSyntax = 'disabled'} = opts;

    if (bundle && typeof opts.runtime === 'string') {
        throw new TypeError('Option `runtime` should be record when `bundle` is enabled.');
    }

    const runtime: RuntimeObj =
        typeof opts?.runtime === 'object'
            ? opts.runtime
            : {style: opts?.runtime ?? '_assets/file-extension.css'};

    const plugin: MarkdownIt.PluginWithOptions<PluginOptions> = (
        md,
        {output = '.', fileExtraAttrs} = {},
    ) => {
        registerTransform(md, {
            bundle,
            runtime,
            output,
            directiveSyntax,
            onBundle: opts.onBundle,
            extraAttrs: fileExtraAttrs ?? opts.extraAttrs,
        });
    };

    Object.assign(plugin, {
        collect(input: string, {destRoot = '.'}: InputOptions) {
            const md = new MarkdownIt().use((md: MarkdownIt) => {
                registerTransform(md, {
                    bundle,
                    runtime,
                    directiveSyntax,
                    output: destRoot,
                    onBundle: opts.onBundle,
                    extraAttrs: opts.extraAttrs,
                });
            });

            md.parse(input, {});
        },
    });

    return plugin;
};

type InputOptions = {
    destRoot: string;
};
