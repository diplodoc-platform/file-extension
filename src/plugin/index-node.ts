import {type RuntimeObj, type TransformOptions, transform as baseTransform} from './transform';

export * from './index';

const onBundle: TransformOptions['onBundle'] = (env, output, runtime) => {
    copyRuntime({runtime, output}, env.bundled);
};

export const transform: typeof baseTransform = (opts) => {
    return baseTransform({onBundle, ...opts});
};

function copyRuntime({runtime, output}: {runtime: RuntimeObj; output: string}, cache: Set<string>) {
    const PATH_TO_RUNTIME = '../runtime';
    const {join, resolve} = require('node:path');
    const runtimeFiles = {
        // 'index.js': runtime.script,
        'index.css': runtime.style,
    };
    for (const [originFile, outputFile] of Object.entries(runtimeFiles)) {
        const file = join(PATH_TO_RUNTIME, originFile);
        if (!cache.has(file)) {
            cache.add(file);
            copy(resolve(__dirname, file), join(output, outputFile));
        }
    }
}

export function copy(from: string, to: string) {
    const {mkdirSync, copyFileSync} = require('node:fs');
    const {dirname} = require('node:path');

    mkdirSync(dirname(to), {recursive: true});
    copyFileSync(from, to);
}
