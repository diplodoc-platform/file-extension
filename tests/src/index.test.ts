import MarkdownIt from 'markdown-it';
import transform from '@diplodoc/transform';

import {type TransformOptions, transform as fileTransformer} from '../../src/plugin';

function html(markup: string, opts?: TransformOptions) {
    return transform(markup, {
        // override the default markdown-it-attrs delimiters,
        // to make it easier to check html for non-valid file markup
        leftDelimiter: '[',
        rightDelimiter: ']',
        plugins: [fileTransformer({bundle: false, ...opts})],
    }).result.html;
}

function meta(markup: string, opts?: TransformOptions) {
    return transform(markup, {
        leftDelimiter: '[',
        rightDelimiter: ']',
        plugins: [fileTransformer({bundle: false, ...opts})],
    }).result.meta;
}

function tokens(markup: string, opts?: TransformOptions) {
    const md = new MarkdownIt().use(fileTransformer({bundle: false, ...opts}));
    return md.parse(markup, {});
}

describe('File extension - plugin', () => {
    it('should render file', () => {
        expect(html('{% file src="../file" name="file.txt" %}')).toMatchSnapshot();
    });

    it('should ignore file markup without params', () => {
        expect(html('{% file %}')).toBe('<p>{% file %}</p>\n');
    });

    it('should not render file without all required attrs', () => {
        expect(html('{% file src="../file" %}')).toBe('<p>{% file src="../file" %}</p>\n');
        expect(html('{% file name="file.txt" %}')).toBe('<p>{% file name="file.txt" %}</p>\n');
    });

    it('should render file with text before', () => {
        expect(html('download it {% file src="../file" name="file.txt" %}')).toMatchSnapshot();
    });

    it('should render file with text after', () => {
        expect(
            html('{% file src="../file" name="file.txt" %} don\'t download it'),
        ).toMatchSnapshot();
    });

    it('should render file between text', () => {
        expect(html('text1 {% file src="../file" name="file.txt" %} 2text')).toMatchSnapshot();
    });

    it('should map all specific file attrs to link html attrs', () => {
        expect(html('{% file src="../file2" name="file2.txt" lang="en" %}')).toMatchSnapshot();
    });

    it('should pass allowed link html attrs', () => {
        expect(
            html(
                '{% file src="../file1" name="file1.txt" referrerpolicy="origin" rel="help" target="_top" type="text/css" %}',
            ),
        ).toMatchSnapshot();
    });

    it('should ignore unknown attrs', () => {
        expect(
            html('{% file src="../file" name="file.txt" foo="1" bar="2" baz="3" %}'),
        ).toMatchSnapshot();
    });

    it('should add extra attrs', () => {
        expect(
            html('{% file src="../file" name="file.txt" %}', {
                extraAttrs: [['data-yfm-file', 'yes']],
            }),
        ).toMatchSnapshot();
    });

    it('should parse attrs with single quotes', () => {
        expect(html("{% file src='index.txt' name='index.html' %}")).toMatchSnapshot();
    });

    it('should render with file markup in attributes', () => {
        expect(
            html('{% file src="in%}dex.txt" name="{% file src=\'a\' name=\'b\' %}" %}'),
        ).toMatchSnapshot();
    });

    it('should render file with different order of attrs', () => {
        expect(
            html('{% file type="text/html" name="page.html" src="../index.html" %}'),
        ).toMatchSnapshot();
    });

    it('should ignore additional special characters', () => {
        expect(html('{% {% file src="index.txt" name="index.html" %} %}')).toMatchSnapshot();
    });

    it('should ignore additional file markup', () => {
        expect(html('{% file {% file src="index.txt" name="index.html" %} %}')).toMatchSnapshot();
    });

    it('should allow quoutes in attribute value', () => {
        expect(html('{% file src="ind\'ex.txt" name=\'ind"ex.html\' %}')).toMatchSnapshot();
    });

    it('should render file with extra spaces around attrs', () => {
        expect(
            html('{% file  src="index.txt"   name="index.html"   type="text/html"  %}'),
        ).toMatchSnapshot();
    });

    it('should not render file without spaces around attrs', () => {
        expect(html('{% file src="index.txt"name="index.html"type="text/html" %}')).toBe(
            `<p>{% file src="index.txt"name="index.html"type="text/html" %}</p>\n`,
        );
    });

    it('should not render file with no spaces near borders', () => {
        expect(html("{%file src='index.txt' name='index.html'%}")).toBe(
            `<p>{%file src='index.txt' name='index.html'%}</p>\n`,
        );
    });

    it('should not add meta if file not parsed', () => {
        expect(meta('parapraph')).toBeUndefined();
    });

    it('should add default style assets to meta', () => {
        expect(meta('{% file src="../file" name="file.txt" %}')).toStrictEqual({
            style: ['_assets/file-extension.css'],
        });
    });

    it('should override style assets name', () => {
        expect(
            meta('{% file src="../file" name="file.txt" %}', {runtime: 'file-ext'}),
        ).toStrictEqual({style: ['file-ext']});
    });

    it('should override style assets name 2', () => {
        expect(
            meta('{% file src="../file" name="file.txt" %}', {runtime: {style: 'yfm-file'}}),
        ).toStrictEqual({style: ['yfm-file']});
    });

    it('should not call onBundle', () => {
        const onBundle = jest.fn(() => {});
        html('text', {bundle: true, onBundle});
        expect(onBundle).not.toHaveBeenCalled();
    });

    it('should call onBundle', () => {
        const onBundle = jest.fn(() => {});
        html('{% file src="../file" name="file.txt" %}', {bundle: true, onBundle});
        expect(onBundle).toHaveBeenCalled();
    });

    it('should generate yfm-file token', () => {
        expect(tokens('{% file src="../file" name="file.txt" %}')).toMatchSnapshot();
    });

    it('should add extra attrs by passing them to plugin options', () => {
        const md = new MarkdownIt();
        const filePlugin = fileTransformer({bundle: false});
        filePlugin(md, {
            fileExtraAttrs: [
                ['data-yfm', '1'],
                ['data-file', '2'],
            ],
        });
        expect(md.render('{% file src="../file" name="file.txt" %}')).toMatchSnapshot();
    });
});
