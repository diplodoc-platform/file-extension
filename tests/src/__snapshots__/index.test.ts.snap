// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File extension - plugin Options: directiveSyntax should render both file markups 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a></p>
<p><a class="yfm-file" href="path/to/video" download="video.mp4"><span class="yfm-file__icon"></span>video.mp4</a></p>
"
`;

exports[`File extension - plugin Options: directiveSyntax should render only new (directive) file markup 1`] = `
"<p>{% file src="../file" name="file.txt" %}</p>
<p><a class="yfm-file" href="path/to/video" download="video.mp4"><span class="yfm-file__icon"></span>video.mp4</a></p>
"
`;

exports[`File extension - plugin Options: directiveSyntax should render only old file markup 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a></p>
<p>:file<a href="path/to/video">video.mp4</a></p>
"
`;

exports[`File extension - plugin dyrective syntax should add allowed attrs 1`] = `
"<p><a class="yfm-file" href="../readme" download="readme.md" referrerpolicy="origin" rel="external" target="_blank" type="text/plain" hreflang="en"><span class="yfm-file__icon"></span>readme.md</a></p>
"
`;

exports[`File extension - plugin dyrective syntax should generate file token 1`] = `
[
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 1,
    "tag": "p",
    "type": "paragraph_open",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": [
      Token {
        "attrs": [
          [
            "class",
            "yfm-file",
          ],
          [
            "href",
            "path/to/video",
          ],
          [
            "download",
            "video.mp4",
          ],
        ],
        "block": false,
        "children": null,
        "content": "video.mp4",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": ":file",
        "meta": null,
        "nesting": 0,
        "tag": "",
        "type": "yfm_file",
      },
    ],
    "content": ":file[video.mp4](path/to/video)",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 0,
    "tag": "",
    "type": "inline",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": null,
    "markup": "",
    "meta": null,
    "nesting": -1,
    "tag": "p",
    "type": "paragraph_close",
  },
]
`;

exports[`File extension - plugin dyrective syntax should not add attrs not from whitelist 1`] = `
"<p><a class="yfm-file" href="a.md" download="a.md"><span class="yfm-file__icon"></span>a.md</a></p>
"
`;

exports[`File extension - plugin dyrective syntax should render file directive 1`] = `
"<p><a class="yfm-file" href="path/to/video" download="video.mp4"><span class="yfm-file__icon"></span>video.mp4</a></p>
"
`;

exports[`File extension - plugin dyrective syntax should render file inside text 1`] = `
"<p>before<a class="yfm-file" href="../../docs/readme.md" download="file.txt"><span class="yfm-file__icon"></span>file.txt</a>after</p>
"
`;

exports[`File extension - plugin should add extra attrs 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file" data-yfm-file="yes"><span class="yfm-file__icon"></span>file.txt</a></p>
"
`;

exports[`File extension - plugin should add extra attrs by passing them to plugin options 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file" data-yfm="1" data-file="2"><span class="yfm-file__icon"></span>file.txt</a></p>
"
`;

exports[`File extension - plugin should allow quoutes in attribute value 1`] = `
"<p><a href="ind'ex.txt" download="ind&quot;ex.html" class="yfm-file"><span class="yfm-file__icon"></span>ind"ex.html</a></p>
"
`;

exports[`File extension - plugin should generate yfm-file token 1`] = `
[
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 1,
    "tag": "p",
    "type": "paragraph_open",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": [
      Token {
        "attrs": [
          [
            "href",
            "../file",
          ],
          [
            "download",
            "file.txt",
          ],
          [
            "class",
            "yfm-file",
          ],
        ],
        "block": false,
        "children": null,
        "content": "file.txt",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": "{% file ",
        "meta": null,
        "nesting": 0,
        "tag": "",
        "type": "yfm_file",
      },
    ],
    "content": "{% file src="../file" name="file.txt" %}",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 0,
    "tag": "",
    "type": "inline",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": null,
    "markup": "",
    "meta": null,
    "nesting": -1,
    "tag": "p",
    "type": "paragraph_close",
  },
]
`;

exports[`File extension - plugin should ignore additional file markup 1`] = `
"<p>{% file <a href="index.txt" download="index.html" class="yfm-file"><span class="yfm-file__icon"></span>index.html</a> %}</p>
"
`;

exports[`File extension - plugin should ignore additional special characters 1`] = `
"<p>{% <a href="index.txt" download="index.html" class="yfm-file"><span class="yfm-file__icon"></span>index.html</a> %}</p>
"
`;

exports[`File extension - plugin should ignore unknown attrs 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a></p>
"
`;

exports[`File extension - plugin should map all specific file attrs to link html attrs 1`] = `
"<p><a href="../file2" download="file2.txt" hreflang="en" class="yfm-file"><span class="yfm-file__icon"></span>file2.txt</a></p>
"
`;

exports[`File extension - plugin should parse attrs with single quotes 1`] = `
"<p><a href="index.txt" download="index.html" class="yfm-file"><span class="yfm-file__icon"></span>index.html</a></p>
"
`;

exports[`File extension - plugin should pass allowed link html attrs 1`] = `
"<p><a href="../file1" download="file1.txt" referrerpolicy="origin" rel="help" target="_top" type="text/css" class="yfm-file"><span class="yfm-file__icon"></span>file1.txt</a></p>
"
`;

exports[`File extension - plugin should render file 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a></p>
"
`;

exports[`File extension - plugin should render file between text 1`] = `
"<p>text1 <a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a> 2text</p>
"
`;

exports[`File extension - plugin should render file with different order of attrs 1`] = `
"<p><a type="text/html" download="page.html" href="../index.html" class="yfm-file"><span class="yfm-file__icon"></span>page.html</a></p>
"
`;

exports[`File extension - plugin should render file with extra spaces around attrs 1`] = `
"<p><a href="index.txt" download="index.html" type="text/html" class="yfm-file"><span class="yfm-file__icon"></span>index.html</a></p>
"
`;

exports[`File extension - plugin should render file with text after 1`] = `
"<p><a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a> don't download it</p>
"
`;

exports[`File extension - plugin should render file with text before 1`] = `
"<p>download it <a href="../file" download="file.txt" class="yfm-file"><span class="yfm-file__icon"></span>file.txt</a></p>
"
`;

exports[`File extension - plugin should render with file markup in attributes 1`] = `
"<p><a href="in%}dex.txt" download="{% file src='a' name='b' %}" class="yfm-file"><span class="yfm-file__icon"></span>{% file src='a' name='b' %}</a></p>
"
`;
